### PR TITLE
Add remove command

### PR DESCRIPTION
**Pull request by darshan3**
Adding a command to remove items from inventory.
### How to test:
1. Give a few pickaxes using "give pickaxe 'amount' "  command
2. Remove one pickaxe using "remove pickaxe" command
3. Remove many pickaxes using "remove pickaxe amount" with various values of the amount
4. Try the remove command when you don't have a pickaxe in the inventory  
5. Try the remove command with amount parameter greater than the amount of pickaxe you have in your inventory. 
  
Test the steps 1 to 5 for blocks, shapes and stackable objects, too.

### CODE REVIEW

**Comment by darshan3** [comment]
Currently, if I request removal of an item with amount `x`, but if the user has only `y` amount of that item where `x`>=`y`, then all `y` items will be removed. Should this be the intended behaviour or should the items be removed only if the amount matches?

**Comment by darshan3** [comment]
@skaldarnar @jellysnake

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 1]
```diff
@@ -114,6 +123,90 @@ public String give(
...
+                            + (shapeUriParam != null ? " (Item can not have a shape)" : "");
+                } else {
+                    return "Could not find "
+                            + (itemAmount > 1 ? itemAmount + " items of " : "an item of ")
+                            + prefab.getName()
+                            + " in your inventory";
+                }
+            }
+        } else if (matches.size() > 1) {
+            StringBuilder builder = new StringBuilder();
```
Comment (jellysnake) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java]: You mentioned that you wanted to try and reduce duplicate code in this file.
You could pull this section and the same in the give command into a private method like "displayItemDuplicates"

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 2]
```diff
@@ -170,4 +263,146 @@ public String bulkGiveItem(
...
+                //Should never be reached
+                return "Block not found";
+            }
+
+            int quantity = quantityParam != null ? quantityParam : 1;
+
+            return removeBlock(blockFamily, quantity, sender);
+
+        } else if (matchingUris.size() > 1) {
+            StringBuilder builder = new StringBuilder();
```
Comment (jellysnake) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java]: Same as my comment in the remove items section. You could move this to a method

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 3]
```diff
@@ -170,4 +263,146 @@ public String bulkGiveItem(
...
+                    if (shapeUriParam == null) {
+                        blockFamily = blockManager.getBlockFamily(new BlockUri(def.get().getUrn(), new ResourceUrn("engine:cube")));
+                    } else {
+                        Set<ResourceUrn> resolvedShapeUris = Assets.resolveAssetUri(shapeUriParam, BlockShape.class);
+                        if (resolvedShapeUris.isEmpty()) {
+                            return "Found block. No shape found for '" + shapeUriParam + "'";
+                        } else if (resolvedShapeUris.size() > 1) {
+                            StringBuilder builder = new StringBuilder();
+                            builder.append("Found block. Non-unique shape name, possible matches: ");
+                            Iterator<ResourceUrn> shapeUris = sortItems(resolvedShapeUris).iterator();
```
Comment (jellysnake) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java]: You could also restructure this and replace it with the method below, as both are `Set<ResourceUrn>`

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 4]
```diff
@@ -170,4 +263,146 @@ public String bulkGiveItem(
...
+            StringBuilder builder = new StringBuilder();
+            builder.append("Non-unique block name, possible matches: ");
+            Joiner.on(", ").appendTo(builder, matchingUris);
+            return builder.toString();
+        }
+
+        return null;
+    }
+
+    private String removeBlock(BlockFamily blockFamily, int itemAmount, EntityRef client) {
```
Comment (jellysnake) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java]: I would prefer a name change here to distinguish it from the prior method. Perhaps something like `doRemoveFromInventory`

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 5]
```diff
@@ -114,6 +123,90 @@ public String give(
...
+        if (itemAmount < 1) {
+            return "Invalid amount of items!";
+        }
+
+        Set<ResourceUrn> matches = assetManager.resolve(itemPrefabName, Prefab.class);
+
+        if (matches.size() == 1) {
+            Prefab prefab = assetManager.getAsset(matches.iterator().next(), Prefab.class).orElse(null);
+
+            if (prefab != null && prefab.getComponent(ItemComponent.class) != null) {
```
Comment (jellysnake) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java]: `hasComponent(foo.class)` should be used over `getComponent(foo.class) != null`

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 6]
```diff
@@ -170,4 +263,146 @@ public String bulkGiveItem(
...
+                    + blockFamily.getDisplayName();
+        } else {
+            return "Could not find "
+                    + (itemAmount > 1 ? itemAmount + " blocks of " : "a block of ")
+                    + blockFamily.getDisplayName()
+                    + " in your inventory";
+        }
+    }
+
+    private <T extends Comparable<T>> List<T> sortItems(Iterable<T> items) {
```
Comment (jellysnake) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java]: Why is this a seperate method?

**Comment by jdrueckert** [comment]
Just tested it out in-game.
1-3 work as expected.

4
- works as expected -> error msg "Could not find an item of Core:pickaxe in your inventory" as no pickaxes in inventory
- would prefer a msg like "Nothing to remove, you don't have Core:pickaxe in your inventory" though

5
- doesn't work as expected -> would expect that as many items as I still have are removed
- e.g. I have 3 pickaxes in my inventory and do `remove pickaxe 7`
- -> `remove 3 pickaxes`
- -> log "Removed 3 Core:pickaxe from your inventory. No more Core:pickaxe items to remove."
- `give pickaxe <amount>` with more than inventory space also gives as many as possible and logs the number it gave. IMO remove should work the same way (reversed).

**Pull request closed without merging**